### PR TITLE
Feature/yta 2945/open rti non rti summary

### PR DESCRIPTION
**Pull request by miloszm**
implementation of charges summary endpoint

### CODE REVIEW

File: app/uk/gov/hmrc/epayeapi/connectors/EpayeConnector.scala [hunk 1]
```diff
@@ -39,21 +41,56 @@ case class EpayeConnector @Inject() (
...
+        .map(q => s"/$q")
+        .getOrElse("")
+
+    get[AnnualSummaryResponse](url, headers)
+  }
 }
 
+object EpayeConnector {
+  def extractTaxYear(taxYear: Option[String]): Option[TaxYear] = {
+    taxYear.flatMap {
```
Comment (thomas-vogel444) [inline app/uk/gov/hmrc/epayeapi/connectors/EpayeConnector.scala]: taxYear should be a string, not an Option[String]

File: app/uk/gov/hmrc/epayeapi/connectors/EpayeConnector.scala [hunk 2]
```diff
@@ -39,21 +41,56 @@ case class EpayeConnector @Inject() (
...
+      case TaxYearPattern(year) => Some(year)
+      case _ => None
+    }
+  }
+
+  object TaxYearPattern {
+    lazy val pattern = """20(\d\d)-(\d\d)""".r
+    def unapply(taxYear: String): Option[TaxYear] = {
+      taxYear match {
```
Comment (thomas-vogel444) [inline app/uk/gov/hmrc/epayeapi/connectors/EpayeConnector.scala]: I would personally get rid of the unapply and put its body in the extractTaxYear directly

File: app/uk/gov/hmrc/epayeapi/connectors/EpayeConnector.scala [hunk 3]
```diff
@@ -39,21 +41,56 @@ case class EpayeConnector @Inject() (
...
+        .map(TaxYear.asString)
+        .map(q => s"/$q")
+        .getOrElse("")
+
+    get[AnnualSummaryResponse](url, headers)
+  }
 }
 
+object EpayeConnector {
+  def extractTaxYear(taxYear: Option[String]): Option[TaxYear] = {
```
[the remark this answers is not in the document]
Comment (thomas-vogel444) [reply app/uk/gov/hmrc/epayeapi/connectors/EpayeConnector.scala]: Also how about putting that method in TaxYear? Looks like an appropriate place :)

File: app/uk/gov/hmrc/epayeapi/services/ChargesSummaryService.scala [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+  def toChargesSummary(annualSummaryResponse: AnnualSummaryResponse): ChargesSummary = {
+    ChargesSummary(
+      rti = toChargesSummaryRti(annualSummaryResponse.rti),
+      non_rti = toChargesSummaryNonRti(annualSummaryResponse.nonRti)
+    )
+  }
+
+  def toChargesSummaryRti(annualSummary: AnnualSummary): Seq[RtiCharge] = {
+    annualSummary.lineItems.map(lineItem =>
```
Comment (thomas-vogel444) [inline app/uk/gov/hmrc/epayeapi/services/ChargesSummaryService.scala]: Call this method rtiChargesSummary?

File: app/uk/gov/hmrc/epayeapi/services/ChargesSummaryService.scala [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        tax_year = TaxYear(lineItem.taxYear.yearFrom, lineItem.taxYear.yearTo),
+        tax_month = lineItem.taxMonth.map(_.month),
+        balance = DebitCredit(lineItem.charges.debit, lineItem.charges.credit),
+        due_date = Some(lineItem.dueDate),
+        is_overdue = lineItem.dueDate.isBefore(today)
+      ))
+  }
+
+  def toChargesSummaryNonRti(annualSummary: AnnualSummary): Seq[NonRtiCharge] = {
+    annualSummary.lineItems.map(lineItem =>
```
Comment (thomas-vogel444) [inline app/uk/gov/hmrc/epayeapi/services/ChargesSummaryService.scala]: Call this method nonRtiChargesSummary?

File: test/uk/gov/hmrc/epayeapi/models/ChargesSummarySpec.scala [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+      val summary = ChargesSummary(
+        rti = Seq(
+          RtiCharge(
+            tax_year = api.TaxYear(2016, 2017),
+            tax_month = Some(3),
+            balance = DebitCredit(debit = 1200.0, credit = 0),
+            due_date = Some(new LocalDate(2016, 10, 22)),
+            is_overdue = true
+          ),
+          RtiCharge(
```
Comment (thomas-vogel444) [inline test/uk/gov/hmrc/epayeapi/models/ChargesSummarySpec.scala]: Use helper factory methods for the test data to reduce noise.

E.g.
```scala
testRtiCharge(taxYear:TaxYear, taxMonth: Option[TaxMonth] = None, debit: Int, credit: Int, dueDate: Option[LocalDate] = None, isOverdue: Boolean = false) = {
 RtiCharge(
             tax_year = taxYear,
             tax_month = taxMonth,
             balance = DebitCredit(debit = debit, credit = credit),
             due_date = dueDate,
             is_overdue = isOverdue
           )
}
```

File: resources/public/api/conf/1.0/examples/ChargesSummary.get.json [hunk 7]
```diff
@@ -0,0 +1,70 @@
+{
```
Comment (thomas-vogel444) [inline resources/public/api/conf/1.0/examples/ChargesSummary.get.json]: I thought we agreed not to use Json for test data but objects/vals

File: app/uk/gov/hmrc/epayeapi/services/ChargesSummaryService.scala (hunk 5 above)
Comment (flashingpumpkin) [reply app/uk/gov/hmrc/epayeapi/services/ChargesSummaryService.scala]: 👍

File: resources/public/api/conf/1.0/examples/ChargesSummary.get.json (hunk 7 above)
Comment (flashingpumpkin) [reply resources/public/api/conf/1.0/examples/ChargesSummary.get.json]: 👍

**Pull request merged**